src/translate_ligo.coffee

```diff
@@ -688,34 +700,90 @@ walk = (root, ctx)->
+            return "var #{config.op_list} : list(operation) := list transaction((#{arg_list.join ' * '}), 0mutez, (get_contract(match_action.callbackAddress) : contract(#{type_str}))) end"
+          
+          when "@respond_append"
+            type_list = []
+            for v in root.arg_list
+              type_list.push translate_type v.type, ctx
+            type_str = type_list.join " * "
             perr "CRITICAL WARNING we don't check balance in send function. So runtime error will be ignored and no throw"
-            # p root
-            return "var #{config.op_list} : list(operation) := list transaction((#{arg_list.join ' * '}), 0mutez, #{config.receiver_name}) end"
+            return "var #{config.op_list} : list(operation) := cons(#{arg_list[0]}, list transaction((#{arg_list[1..].join ' * '}), 0mutez, (get_contract(match_action.callbackAddress) : contract(#{type_str})) end)"
+          
           else
             fn = root.fn.name
       else
         fn = walk root.fn, ctx
-    
+      
       if arg_list.length == 0
         arg_list.push "unit"
       
-      ret_types_list = []
-      # type can be null
-      # type can be contract name, so no nest_list
-      return_types = root.fn.type?.nest_list[1]
-      for v in return_types?.nest_list or []
-        ret_types_list.push translate_type v, ctx
       
-      tmp_var = "tmp_#{ctx.tmp_idx++}"
       call_expr = "#{fn}(#{arg_list.join ', '})";
-
-      if not root.left_unpack
-        "#{call_expr}"
+      
+      if not root.left_unpack or fn in ["get_contract", "transaction"]
+        call_expr
       else
-        if ret_types_list.length == 1
-          ctx.sink_list.push "const #{tmp_var} : #{ret_types_list[0]} = #{call_expr}"
+        if root.fn_decl
+          {
+            returns_op_list
+            modifies_storage
+            returns_value
+          } = root.fn_decl
+          {type_o} = root.fn_decl
+        else if type_decl = ti_map[root.fn.name]
+          returns_op_list = false
+          modifies_storage= false
+          returns_value   = type_decl.nest_list[1].nest_list.length > 0
+          type_o          = type_decl.nest_list[1]
         else
-          ctx.sink_list.push "const #{tmp_var} : (#{ret_types_list.join ' * '}) = #{call_expr}"
+          perr "WARNING !root.fn_decl #{root.fn.name}"
+          return call_expr
+        
+        ret_types_list = []
+        for v in type_o.nest_list
+          ret_types_list.push translate_type v, ctx
+        
+        if ret_types_list.length == 0
+          call_expr
+        else if ret_types_list.length == 1 and returns_value
+          ctx.terminate_expr_replace_fn = ()->
+            perr "WARNING #{call_expr} was terminated with dummy var_decl"
+            tmp_var = "terminate_tmp_#{ctx.tmp_idx++}"
+            "const #{tmp_var} : (#{ret_types_list.join ' * '}) = #{call_expr}"
+          ctx.terminate_expr_check = call_expr
+        else
```

Comment:
I believe this should be its own ast transform which assigns function calls to either `ops` or `state`

Reply:
But this logic is used only here, so we would not reuse it anywhere else

Reply:
It is not about reusing.
I believe code will look structurally better and be easier to understand if we would move complex logic from `translate_ligo` to transforms